test: restart coordinator during topology operation

aleksbykov opened the pull request:
If coordinator node was killed, restarted, become not
operatable during topology operation, new coordinator should be elected,
operation should be aborted and cluster should be rolled back

Test is monitoring log message on target node where topology
operation is running  and reboot coordinator after
required log message appeared.

mykaul commented on test/topology_experimental_raft/test_coordinator_aborting.py:
```diff
@@ -0,0 +1,162 @@
+async def get_coordinator_id(manager):
+    """ Get coordintor host id from history
+
+    Select all desciription where new coordinator was elecected
```
```suggestion
    Select all description where new coordinator was elected
```

kbr-scylla commented on test/topology_experimental_raft/test_coordinator_aborting.py:
```diff
@@ -0,0 +1,163 @@
+logger = logging.getLogger(__name__)
+
+
+async def restart_coordinator(manager: ManagerClient, expr, coordinator: ServerInfo, 
```
`expr: str`

kbr-scylla commented on test/topology_experimental_raft/test_coordinator_aborting.py:
```diff
@@ -0,0 +1,163 @@
+
+@pytest.mark.asyncio
+async def test_kill_coordinator_during_op(manager: ManagerClient) -> None:
+    """ Kill cooridinator while topology operation running for cluster: decommission,
```
space at the beginning of comment should be removed
`cooridinator` -> `coordinator`
`running` -> `is running`

kbr-scylla commented on test/topology_experimental_raft/test_coordinator_aborting.py:
```diff
@@ -0,0 +1,163 @@
+    bootstrap, removenode, replace.
+
+    1. Start operation on target node
+    2. wait appropriate message in log on target node
```
`wait` -> `wait for`

kbr-scylla commented on test/topology_experimental_raft/test_coordinator_aborting.py:
```diff
@@ -0,0 +1,163 @@
+
+    1. Start operation on target node
+    2. wait appropriate message in log on target node
+    3. Stop coordinator node, after target node reach requiered point
```
`reach` -> `reaches`
`requiered` -> `required`

kbr-scylla reviewed (changes requested):
Please make sure to use proper English grammar and avoid spelling mistakes. These errors are very distracting and take away attention from the important stuff. Unfortunately this PR is full of them, almost every sentence has a mistake.

Whenever sending a PR to scylladb repo, please do a self-review first, correcting these errors.

kbr-scylla commented on test/topology_experimental_raft/test_coordinator_aborting.py:
```diff
@@ -0,0 +1,183 @@
+
+    await asyncio.gather(manager.decommission_node(server_id=other_nodes[-1].server_id,
+                                                                     expected_error="Decommission failed. See earlier errors"),
+                         restart_coordinator_after_log_message_on_target_node(manager, 
```
Did anything change here? This is still susceptible to races, and won't work every time.

You're enforcing an order between these two events:
E1: target node starts decommission (by printing "request decommission for ...")
E2: coordinator is restarted

We can write this in short: E1 < E2.

But there is nothing that guarantees when decommission will end or fail.
Let's say that
E3: decommission ends

The test does not do anything to ensure that E2 < E3. It could happen that decommission ends before coordinator is restarted.

And even if you had:
E1 < E2 < E3

it would still not be enough: coordinator could get restarted at a point where decommission no longer needs coordinator to succeed, i.e. coordinator has finished its work, and sends message to target node telling it that decommission has succeeded, and then the test restarts it. In that case decommission still won't fail.

You need to somehow guarantee that coordinator is restarted **after decommission starts**, but **before target node stops needing the coordinator node**.

One way to ensure that, which comes to my mind, is to have an injection in the target node's code that will pause the node when coordinator is waiting for it. For example, when coordinator sends stream request to node and waits for it to finish streaming. Inside streaming code you could have an injection that pauses the node and waits for message from test. The injection would print a log message then pause the node. The test would wait for this message, then kill the coordinator, then send message to target node to unpause it. This would enforce an order of the following events:
E1: target node starts decommission (by printing "request decommission for ...")
E2: coordinator is restarted
E3: target node sends streaming finished message to coordinator

E1 < E2 < E3

Then we could expect decommission to fail.

kbr-scylla commented:
It is customary in Scylla core repository to post a summary of what changed since the last version whenever you send an update to the PR. See here for example: https://github.com/scylladb/scylladb/pull/16192#issuecomment-1833952377

aleksbykov commented:
new changes:
 - error injection to topology_class before stream began in storage_service.cc
 - new test with error injection instead of waiting log message

aleksbykov commented:
New changes:
 - Removed tests based on waiting log messages
 - Left 2 stable cases and added to test description the table with operation coverage

mykaul commented on test/topology_experimental_raft/test_crash_coordinator_before_streaming.py:
```diff
@@ -0,0 +1,131 @@
+
+    
+
+async def get_coordinator_host_ids(manager: ManagerClient) -> list[str]:
```
The following 3 functions look like they can be useful outside this test, no?

aleksbykov replied:
done

kbr-scylla commented on test/topology_experimental_raft/test_crash_coordinator_before_streaming.py:
```diff
@@ -0,0 +1,131 @@
+    await manager.decommission_node(server_id=other_nodes[-1].server_id, expected_error="Decommission failed. See earlier errors")
+
+    coordinators_ids = await get_coordinator_host_ids(manager)
+    assert len(coordinators_ids) == 2, "No new record about the coordinator's election"
```
There is no guarantee how fast the new coordinator will be elected. It should generally take less than 2 seconds, but we cannot depend on such timing.

Now the test will race with Scylla and sometimes you will reach this check before Scylla manages to elect the new coordinator.

I suggest using `wait_for` with reasonable timeout (60 seconds?) to wait for the condition that a new coordinator was elected.

kbr-scylla commented on test/topology_experimental_raft/test_crash_coordinator_before_streaming.py:
```diff
@@ -0,0 +1,131 @@
+                              expected_error="Removenode failed. See earlier errors")
+    
+    coordinators_ids = await get_coordinator_host_ids(manager)
+    assert len(coordinators_ids) == 3, "No new record about the coordinator's election"
```
Same as above.

aleksbykov commented:
@kbr-scylla can you take a look?

kbr-scylla reviewed (approved):
Why draft? I think it's time to undraft the PR.

Also please run the test in debug mode too, at least 50 times.
To speed it up you can increase concurrency of those runs by temporarily setting `pool_size` in test/topology_experimental_raft/suite.yaml to a higher value like `10` (but be careful not to kill your machine)

aleksbykov commented:
test are passed with debug mode:
```
$ ./tools/toolchain/dbuild ./test.py --repeat=50 topology_experimental_raft/test_crash_coordinator_before_streaming
Found 50 tests.
================================================================================
[N/TOTAL]   SUITE    MODE   RESULT   TEST
------------------------------------------------------------------------------
[1/50]     topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.2 153.33s
[2/50]     topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.1 157.66s
[3/50]     topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.3 139.85s
[4/50]     topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.4 146.87s
[5/50]     topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.5 143.73s
[6/50]     topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.6 142.78s
[7/50]     topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.7 152.31s
[8/50]     topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.8 151.69s
[9/50]     topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.9 140.05s
[10/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.10 138.80s
[11/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.11 138.55s
[12/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.12 150.40s
[13/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.13 153.27s
[14/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.14 137.81s
[15/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.15 143.18s
[16/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.16 142.93s
[17/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.17 146.21s
[18/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.18 155.50s
[19/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.19 146.67s
[20/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.20 161.55s
[21/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.21 155.68s
[22/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.22 154.92s
[23/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.23 151.74s
[24/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.24 136.33s
[25/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.25 153.21s
[26/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.26 153.08s
[27/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.27 139.58s
[28/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.28 139.09s
[29/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.29 140.20s
[30/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.30 158.64s
[31/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.31 151.69s
[32/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.32 143.46s
[33/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.33 162.20s
[34/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.34 149.85s
[35/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.35 164.16s
[36/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.36 153.97s
[37/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.38 151.36s
[38/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.37 158.36s
[39/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.40 139.82s
[40/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.39 155.30s
[41/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.41 144.03s
[42/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.42 161.68s
[43/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.43 158.23s
[44/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.44 149.71s
[45/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.45 139.67s
[46/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.46 150.36s
[47/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.47 149.13s
[48/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.48 162.00s
[49/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.49 157.13s
[50/50]    topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.50 150.71s
------------------------------------------------------------------------------
CPU utilization: 26.7%

```

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6144/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6144/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7445/)
#### Failed Tests (8/24125):
- [topology.test_coordinator_queue_management.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6144/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_coordinator_queue_management_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_coordinator_queue_management.dev.2)
- [topology_experimental_raft.test_crash_coordinator_before_streaming.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6144/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_crash_coordinator_before_streaming_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_crash_coordinator_before_streaming.release.1)
- [topology_experimental_raft.test_crash_coordinator_before_streaming.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6144/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_crash_coordinator_before_streaming_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_crash_coordinator_before_streaming.release.2)
- [topology_experimental_raft.test_crash_coordinator_before_streaming.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6144/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_crash_coordinator_before_streaming_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_crash_coordinator_before_streaming.release.3)
- [test_coordinator_queue_management](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6144/testReport/junit/%28root%29/test_coordinator_queue_management/Tests___Unit_Tests___test_coordinator_queue_management) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_coordinator_queue_management)
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6144/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/Tests___Unit_Tests___test_kill_coordinator_during_op) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6144/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/Tests___Unit_Tests___test_kill_coordinator_during_op_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6144/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/Tests___Unit_Tests___test_kill_coordinator_during_op_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
#### Build Details:
- Duration: 2 hr 57 min
- Builder: spider1.cloudius-systems.com

aleksbykov commented:
update: do not run test in release mode

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6168/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6168/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7468/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6168/)
#### Failed Tests (2/24120):
- [test_tablet_cleanup](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6168/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_cleanup) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_cleanup)
- [topology_experimental_raft.test_tablets.dev.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6168/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_dev_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.dev.4)
#### Build Details:
- Duration: 1 hr 52 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6184/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6184/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7486/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6184/)
#### Build Details:
- Duration: 1 hr 54 min
- Builder: spider8.cloudius-systems.com

aleksbykov commented on test/topology_experimental_raft/test_crash_coordinator_before_streaming.py:
```diff
@@ -0,0 +1,169 @@
+    return list of coordinator host ids, where
+    first element in list is active coordinator
+    """
+    stm = SimpleStatement("select description from system.group0_history \
```
fixed

aleksbykov commented on test/topology_experimental_raft/test_crash_coordinator_before_streaming.py:
```diff
@@ -0,0 +1,169 @@
+    return uuid
+
+
+async def wait_new_coordinator_elected(manager: ManagerClient, expected_num_of_elections: int, deadline: float):
```
fixed

aleksbykov commented on test/topology_experimental_raft/test_crash_coordinator_before_streaming.py:
```diff
@@ -0,0 +1,169 @@
+
+def get_uuid_from_str(string: str) -> str:
+    """Search uuid in string"""
+    uuid_regex = re.compile(r"([0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12})")
```
set it as module var

aleksbykov commented on test/topology_experimental_raft/test_crash_coordinator_before_streaming.py:
```diff
@@ -0,0 +1,169 @@
+    coordinators_ids = await get_coordinator_host_ids(manager)
+    assert len(coordinators_ids) == 1, "At least 1 coordinator id should be found"
+
+    # kill coordinator during decommission
```
added also log message

aleksbykov commented on test/topology_experimental_raft/test_crash_coordinator_before_streaming.py:
```diff
@@ -0,0 +1,169 @@
+import logging
+import re
+import time
+import pytest
```
done

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6249/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6249/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7549/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6249/)
#### Build Details:
- Duration: 1 hr 56 min
- Builder: spider2.cloudius-systems.com

kbr-scylla commented:
Queued

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7079/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7079/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/214/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8254/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7079/)
#### Failed Tests (12/23465):
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7079/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/test_kill_coordinator_during_op) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7079/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/test_kill_coordinator_during_op_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7079/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/test_kill_coordinator_during_op_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7079/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/test_kill_coordinator_during_op_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7079/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/test_kill_coordinator_during_op_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7079/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/test_kill_coordinator_during_op_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [topology_experimental_raft.test_crash_coordinator_before_streaming.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7079/testReport/junit/%28root%29/non-boost%20tests/topology_experimental_raft_test_crash_coordinator_before_streaming_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_crash_coordinator_before_streaming.debug.1)
- [topology_experimental_raft.test_crash_coordinator_before_streaming.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7079/testReport/junit/%28root%29/non-boost%20tests/topology_experimental_raft_test_crash_coordinator_before_streaming_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_crash_coordinator_before_streaming.debug.2)
- [topology_experimental_raft.test_crash_coordinator_before_streaming.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7079/testReport/junit/%28root%29/non-boost%20tests/topology_experimental_raft_test_crash_coordinator_before_streaming_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_crash_coordinator_before_streaming.debug.3)
- [topology_experimental_raft.test_crash_coordinator_before_streaming.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7079/testReport/junit/%28root%29/non-boost%20tests/topology_experimental_raft_test_crash_coordinator_before_streaming_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_crash_coordinator_before_streaming.dev.1)
#### Build Details:
- Duration: 1 hr 59 min
- Builder: spider1.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7113/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7113/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/250/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8288/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7113/)
#### Failed Tests (12/23483):
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7113/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/test_kill_coordinator_during_op) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7113/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/test_kill_coordinator_during_op_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7113/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/test_kill_coordinator_during_op_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7113/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/test_kill_coordinator_during_op_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7113/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/test_kill_coordinator_during_op_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7113/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/test_kill_coordinator_during_op_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [topology_experimental_raft.test_crash_coordinator_before_streaming.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7113/testReport/junit/%28root%29/non-boost%20tests/topology_experimental_raft_test_crash_coordinator_before_streaming_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_crash_coordinator_before_streaming.debug.1)
- [topology_experimental_raft.test_crash_coordinator_before_streaming.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7113/testReport/junit/%28root%29/non-boost%20tests/topology_experimental_raft_test_crash_coordinator_before_streaming_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_crash_coordinator_before_streaming.debug.2)
- [topology_experimental_raft.test_crash_coordinator_before_streaming.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7113/testReport/junit/%28root%29/non-boost%20tests/topology_experimental_raft_test_crash_coordinator_before_streaming_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_crash_coordinator_before_streaming.debug.3)
- [topology_experimental_raft.test_crash_coordinator_before_streaming.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7113/testReport/junit/%28root%29/non-boost%20tests/topology_experimental_raft_test_crash_coordinator_before_streaming_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_crash_coordinator_before_streaming.dev.1)
#### Build Details:
- Duration: 2 hr 21 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7125/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7125/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/268/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8306/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7125/)
#### Failed Tests (12/23483):
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7125/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/test_kill_coordinator_during_op) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7125/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/test_kill_coordinator_during_op_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7125/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/test_kill_coordinator_during_op_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7125/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/test_kill_coordinator_during_op_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7125/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/test_kill_coordinator_during_op_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7125/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/test_kill_coordinator_during_op_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [topology_experimental_raft.test_crash_coordinator_before_streaming.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7125/testReport/junit/%28root%29/non-boost%20tests/topology_experimental_raft_test_crash_coordinator_before_streaming_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_crash_coordinator_before_streaming.debug.1)
- [topology_experimental_raft.test_crash_coordinator_before_streaming.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7125/testReport/junit/%28root%29/non-boost%20tests/topology_experimental_raft_test_crash_coordinator_before_streaming_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_crash_coordinator_before_streaming.debug.2)
- [topology_experimental_raft.test_crash_coordinator_before_streaming.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7125/testReport/junit/%28root%29/non-boost%20tests/topology_experimental_raft_test_crash_coordinator_before_streaming_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_crash_coordinator_before_streaming.debug.3)
- [topology_experimental_raft.test_crash_coordinator_before_streaming.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7125/testReport/junit/%28root%29/non-boost%20tests/topology_experimental_raft_test_crash_coordinator_before_streaming_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_crash_coordinator_before_streaming.dev.1)
#### Build Details:
- Duration: 1 hr 45 min
- Builder: spider3.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7130/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7130/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8311/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/273/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7130/)
#### Failed Tests (2/23483):
- [test_remove_garbage_group0_members](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7130/testReport/junit/%28root%29/test_topology_remove_garbage_group0/test_remove_garbage_group0_members) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_remove_garbage_group0_members)
- [topology_custom.test_topology_remove_garbage_group0.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7130/testReport/junit/%28root%29/non-boost%20tests/topology_custom_test_topology_remove_garbage_group0_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_topology_remove_garbage_group0.debug.3)
#### Build Details:
- Duration: 2 hr 53 min
- Builder: i-0a6e9f13f5bbee6f4 (m5d.12xlarge)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7191/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7191/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/332/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8370/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7191/)
#### Build Details:
- Duration: 2 hr 13 min
- Builder: spider5.cloudius-systems.com

aleksbykov commented:
@kbr-scylla , can you take a look on it?
new:
Removed cycle in search coordinator node
Removed node after fail and restore cluster to initial size

Arm also passed in debug mode: https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/2112/testReport/(root)/non-boost%20tests/ 50 times

piodul reviewed (commented):
Please rebase.

I asked @kbr-scylla to review, he said he will try to review it tomorrow.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7932/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7932/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/900/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8928/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7932/)
#### Failed Tests (1/30261):
- [test_concurrent_restart](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7932/testReport/junit/nodetool_additional_test/TestNodetool/Tests___dtest___test_concurrent_restart) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_concurrent_restart)
#### Build Details:
- Duration: 2 hr 40 min
- Builder: spider5.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7938/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7938/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8932/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/904/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7938/)
#### Failed Tests (2/30277):
- [test_fence_hints](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7938/testReport/junit/%28root%29/test_fencing/Tests___Unit_Tests___test_fence_hints) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_fence_hints)
- [topology_experimental_raft.test_fencing.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7938/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_fencing_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_fencing.debug.1)
#### Build Details:
- Duration: 2 hr 27 min
- Builder: spider1.cloudius-systems.com

kbr-scylla commented on test/topology/util.py:
```diff
@@ -327,3 +330,62 @@ async def trigger_snapshot(manager, server: ServerInfo) -> None:
+        coordinator_host_id = get_uuid_from_str(row.description)
+        if coordinator_host_id:
+            coordinators_ids.append(coordinator_host_id)
+        continue
```
nit: redundant `continue`

aleksbykov replied:
fixe

kbr-scylla commented on test/topology/util.py:
```diff
@@ -327,3 +330,62 @@ async def trigger_snapshot(manager, server: ServerInfo) -> None:
+        if len(coordinators_ids) == expected_num_of_elections \
+            and coordinators_ids[0] != coordinators_ids[1]:
+            return True
+        logger.error("New coordinator was not elected %s", coordinators_ids)
```
This is not an error. It's only an error if we timeout. But we may need to wait for some time until coordinator is elected which is expected.

So you could make it `warn` or `info`.

aleksbykov replied:
set as warning

kbr-scylla commented on test/topology_experimental_raft/suite.yaml:
```diff
@@ -13,6 +13,7 @@ run_first:
   - test_tablets
 skip_in_release:
   - test_raft_cluster_features
+  - test_crash_coordinator_before_streaming
```
Use the new `skip_mode` marker instead (you can find examples in other tests, grep for them)

aleksbykov replied:
used

kbr-scylla reviewed (commented):
Only small comments.

Please run the latest version of the test 100 times on both architectures in all applicable modes using the byo job.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7990/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7990/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8986/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/958/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7990/)
#### Failed Tests (2/30184):
- [test_kill_coordinator_during_op](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7990/testReport/junit/%28root%29/test_crash_coordinator_before_streaming/Tests___Unit_Tests___test_kill_coordinator_during_op) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_kill_coordinator_during_op)
- [topology_experimental_raft.test_crash_coordinator_before_streaming.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7990/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_crash_coordinator_before_streaming_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_crash_coordinator_before_streaming.dev.1)
#### Build Details:
- Duration: 2 hr 30 min
- Builder: spider8.cloudius-systems.com

kbr-scylla commented:
@aleksbykov I looked at the last failure, it looks like a bug in Scylla, in particular gossiper failure detector.

Failure happened here:
```
        # kill coordinator during replace
        logger.debug("Kill coordinator during replace")
        nodes = await manager.running_servers()
        coordinator_host = await get_coordinator_host(manager)
        other_nodes = [srv for srv in nodes if srv.server_id != coordinator_host.server_id]
        node_to_replace_srv_id = other_nodes[-1].server_id
        await manager.server_stop_gracefully(node_to_replace_srv_id)
        await manager.api.enable_injection(coordinator_host.ip_addr, "crash_coordinator_before_stream", one_shot=True)
        replace_cfg = ReplaceConfig(replaced_id = node_to_replace_srv_id, reuse_ip_addr = False, use_host_id = True)
        new_node = await manager.server_add(start=False, replace_cfg=replace_cfg)
        await manager.server_start(new_node.server_id, expected_error="Replace failed. See earlier errors")
        await wait_new_coordinator_elected(manager, 5, time.time() + 60)
        logger.debug("Start old coordinator node")
        await manager.server_restart(coordinator_host.server_id)
        await manager.servers_see_each_other(await manager.running_servers())
        logger.debug("Replaced node is already non-voter and will be banned after restart. Remove it")
        coordinator_host = await get_coordinator_host(manager)
>       await manager.remove_node(coordinator_host.server_id, node_to_replace_srv_id)
```

Failure happened because one of the server considered the coordinator as DOWN during repair, even though we restarted it and waited for servers to see each other as alive (`await manager.servers_see_each_other(await manager.running_servers())`):
```
https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7990/artifact/testlog/x86_64/dev/scylla-3132.log
```
ERROR 2024-04-15 18:03:27,869 [shard 0:strm] repair - repair[f71c709c-3090-435d-8e79-59cc253beafe]: Repair 1 out of 28 ranges, keyspace=system_distributed, table=cdc_streams_descriptions_v2, range=(-2861714195313478996,-2763826702956306527], peers={127.18.81.53, 127.18.81.45}, live_peers={127.18.81.53}, status=failed: mandatory neighbor=127.18.81.45 is not alive
```

`127.18.81.45` is the old coordinator that we killed and restarted.

We can see ealier in the logs of `scylla-3132` that it marked the node as alive when it restarted, but then for some reason, marked it as DOWN again. It looks like some kind of race in gossiper failure detector logic:
```
ERROR 2024-04-15 18:03:23,747 [shard 0:main] rpc - client 127.18.81.45:7000: client connection dropped: recv: Connection reset by peer
ERROR 2024-04-15 18:03:23,748 [shard 0:main] rpc - client 127.18.81.45:7000: client connection dropped: recv: Connection reset by peer
WARN  2024-04-15 18:03:23,765 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.18.81.45, status = failed: seastar::rpc::closed_error (connection is closed)
WARN  2024-04-15 18:03:23,765 [shard 1:strm] gossip - failure_detector_loop: Send echo to node 127.18.81.45, status = failed: seastar::rpc::closed_error (connection is closed)
INFO  2024-04-15 18:03:24,448 [shard 0:main] raft_group_registry - marking Raft server e44f5621-6925-4b8c-9976-de937c70961d as dead for raft groups
WARN  2024-04-15 18:03:25,765 [shard 1:strm] gossip - failure_detector_loop: Send echo to node 127.18.81.45, status = failed: seastar::rpc::closed_error (connection is closed)
DEBUG 2024-04-15 18:03:25,893 [shard 0:strm] raft_topology - reload raft topology state
DEBUG 2024-04-15 18:03:25,894 [shard 0: gms] raft_topology - topology cmd rpc barrier_and_drain is called
INFO  2024-04-15 18:03:25,895 [shard 0:strm] token_metadata - Added node a6727b22-60a6-48c3-bbba-15a5ab9c66c3 as pending replacing endpoint which replaces existing node 6acbc264-e4af-4687-9a69-f602867506f9
DEBUG 2024-04-15 18:03:25,896 [shard 0:strm] raft_topology - topology_state_load: the last committed CDC generation ID: (2024/04/15 15:03:14, 48b203fa-fb39-11ee-2792-279e47a599d9)
DEBUG 2024-04-15 18:03:25,896 [shard 0: gms] raft_topology - Got raft_topology_cmd::barrier_and_drain, version 29, current version 29
DEBUG 2024-04-15 18:03:25,896 [shard 0: gms] raft_topology - raft_topology_cmd::barrier_and_drain done
DEBUG 2024-04-15 18:03:25,896 [shard 1: gms] raft_topology - Got raft_topology_cmd::barrier_and_drain, version 29, current version 29
DEBUG 2024-04-15 18:03:25,896 [shard 1: gms] raft_topology - raft_topology_cmd::barrier_and_drain done
DEBUG 2024-04-15 18:03:25,899 [shard 0: gms] raft_topology - topology cmd rpc barrier is called
DEBUG 2024-04-15 18:03:25,899 [shard 0:strm] raft_topology - reload raft topology state
INFO  2024-04-15 18:03:25,900 [shard 0:strm] token_metadata - Added node a6727b22-60a6-48c3-bbba-15a5ab9c66c3 as pending replacing endpoint which replaces existing node 6acbc264-e4af-4687-9a69-f602867506f9
DEBUG 2024-04-15 18:03:25,901 [shard 0:strm] raft_topology - topology_state_load: the last committed CDC generation ID: (2024/04/15 15:03:14, 48b203fa-fb39-11ee-2792-279e47a599d9)
DEBUG 2024-04-15 18:03:25,905 [shard 0: gms] raft_topology - topology cmd rpc barrier_and_drain is called
DEBUG 2024-04-15 18:03:25,905 [shard 0:strm] raft_topology - reload raft topology state
INFO  2024-04-15 18:03:25,906 [shard 0:strm] token_metadata - Added node a6727b22-60a6-48c3-bbba-15a5ab9c66c3 as pending replacing endpoint which replaces existing node 6acbc264-e4af-4687-9a69-f602867506f9
DEBUG 2024-04-15 18:03:25,907 [shard 0:strm] raft_topology - topology_state_load: the last committed CDC generation ID: (2024/04/15 15:03:14, 48b203fa-fb39-11ee-2792-279e47a599d9)
DEBUG 2024-04-15 18:03:25,907 [shard 0: gms] raft_topology - Got raft_topology_cmd::barrier_and_drain, version 30, current version 30
DEBUG 2024-04-15 18:03:25,907 [shard 0: gms] raft_topology - raft_topology_cmd::barrier_and_drain done
DEBUG 2024-04-15 18:03:25,907 [shard 1: gms] raft_topology - Got raft_topology_cmd::barrier_and_drain, version 30, current version 30
DEBUG 2024-04-15 18:03:25,907 [shard 1: gms] raft_topology - raft_topology_cmd::barrier_and_drain done
DEBUG 2024-04-15 18:03:25,912 [shard 0:strm] raft_topology - reload raft topology state
DEBUG 2024-04-15 18:03:25,912 [shard 0: gms] raft_topology - topology cmd rpc barrier is called
INFO  2024-04-15 18:03:25,913 [shard 0:strm] token_metadata - Added node a6727b22-60a6-48c3-bbba-15a5ab9c66c3 as pending replacing endpoint which replaces existing node 6acbc264-e4af-4687-9a69-f602867506f9
DEBUG 2024-04-15 18:03:25,914 [shard 0:strm] raft_topology - topology_state_load: the last committed CDC generation ID: (2024/04/15 15:03:14, 48b203fa-fb39-11ee-2792-279e47a599d9)
DEBUG 2024-04-15 18:03:25,919 [shard 0:strm] raft_topology - reload raft topology state
INFO  2024-04-15 18:03:25,920 [shard 0:strm] token_metadata - Added node a6727b22-60a6-48c3-bbba-15a5ab9c66c3 as pending replacing endpoint which replaces existing node 6acbc264-e4af-4687-9a69-f602867506f9
DEBUG 2024-04-15 18:03:25,921 [shard 0:strm] raft_topology - topology_state_load: the last committed CDC generation ID: (2024/04/15 15:03:14, 48b203fa-fb39-11ee-2792-279e47a599d9)
INFO  2024-04-15 18:03:25,993 [shard 0:main] raft_group_registry - marking Raft server a6727b22-60a6-48c3-bbba-15a5ab9c66c3 as dead for raft groups
DEBUG 2024-04-15 18:03:25,997 [shard 0:strm] raft_topology - reload raft topology state
INFO  2024-04-15 18:03:25,998 [shard 0:strm] gossip - Removed endpoint 127.18.81.8
INFO  2024-04-15 18:03:25,998 [shard 0:strm] gossip - InetAddress a6727b22-60a6-48c3-bbba-15a5ab9c66c3/127.18.81.8 is now DOWN, status = UNKNOWN
INFO  2024-04-15 18:03:25,998 [shard 0:strm] gossip - Finished to force remove node 127.18.81.8
DEBUG 2024-04-15 18:03:26,000 [shard 0:strm] raft_topology - topology_state_load: the last committed CDC generation ID: (2024/04/15 15:03:14, 48b203fa-fb39-11ee-2792-279e47a599d9)
INFO  2024-04-15 18:03:26,748 [shard 0:main] raft_group_registry - marking Raft server e44f5621-6925-4b8c-9976-de937c70961d as alive for raft groups
INFO  2024-04-15 18:03:27,662 [shard 0: gms] gossip - Node 127.18.81.45 has restarted, now UP, status = NORMAL
WARN  2024-04-15 18:03:27,765 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.18.81.8, status = failed: seastar::rpc::closed_error (connection is closed)
INFO  2024-04-15 18:03:27,766 [shard 1:strm] gossip - failure_detector_loop: Mark node 127.18.81.45 as DOWN
INFO  2024-04-15 18:03:27,766 [shard 0:strm] gossip - InetAddress e44f5621-6925-4b8c-9976-de937c70961d/127.18.81.45 is now DOWN, status = NORMAL
```

In particular:
```
WARN  2024-04-15 18:03:25,765 [shard 1:strm] gossip - failure_detector_loop: Send echo to node 127.18.81.45, status = failed: seastar::rpc::closed_error (connection is closed)
...
INFO  2024-04-15 18:03:27,662 [shard 0: gms] gossip - Node 127.18.81.45 has restarted, now UP, status = NORMAL
...
INFO  2024-04-15 18:03:27,766 [shard 1:strm] gossip - failure_detector_loop: Mark node 127.18.81.45 as DOWN
```

Hm. Perhaps this `closed_error` happened because the node was killed, and then triggered the "Mark as DOWN", but this raced with node being marked as UP in the middle?

cc @piodul --- could this be related somehow to 2d9e78b09af14c3a2acbd3af7c8910e514895d36?

aleksbykov commented:
@kbr-scylla 
thanks, i am trying to investigate that issue right now. it didn't reproduced always and only with dev mode (in debug mode it is passed always).
I have another branch, and if old coordinator was restart second time, then test always passed

kbr-scylla commented:
Uploaded logs from last failure:

[scylla-3131.log](https://github.com/scylladb/scylladb/files/14996001/scylla-3131.log)
[scylla-3132.log](https://github.com/scylladb/scylladb/files/14996002/scylla-3132.log)
[scylla-3133.log](https://github.com/scylladb/scylladb/files/14996003/scylla-3133.log)
[scylla-3136.log](https://github.com/scylladb/scylladb/files/14996004/scylla-3136.log)
[scylla-3137.log](https://github.com/scylladb/scylladb/files/14996005/scylla-3137.log)
[scylla-3154.log](https://github.com/scylladb/scylladb/files/14996006/scylla-3154.log)
[scylla-3156.log](https://github.com/scylladb/scylladb/files/14996007/scylla-3156.log)
[scylla-3168.log](https://github.com/scylladb/scylladb/files/14996008/scylla-3168.log)
[topology_experimental_raft.test_crash_coordinator_before_streaming.1.log](https://github.com/scylladb/scylladb/files/14996009/topology_experimental_raft.test_crash_coordinator_before_streaming.1.log)

aleksbykov commented:
I have reproduced it locally:
```
(scylla-test-py-3.11.4) abykov@scyllabuntu:~/Projects/scylla(kill-coordinator-during-topology-ops)$ ./tools/toolchain/dbuild ./test.py --mode=dev --repeat 3 topology_experimental_raft/test_crash_coordinator_before_streaming
Found 3 tests.
================================================================================
[N/TOTAL]   SUITE    MODE   RESULT   TEST
------------------------------------------------------------------------------
[1/3]      topology_experimental_raft   dev   [ FAIL ] topology_experimental_raft.test_crash_coordinator_before_streaming.2 25.43s
[2/3]      topology_experimental_raft   dev   [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.3 39.87s
[3/3]      topology_experimental_raft   dev   [ PASS ] topology_experimental_raft.test_crash_coordinator_before_streaming.1 40.81s

```

[reproduce-3-runs.tar.gz](https://github.com/scylladb/scylladb/files/14996051/reproduce-3-runs.tar.gz)

kbr-scylla commented:
@aleksbykov did you change the test in that last push?

aleksbykov commented:
> @aleksbykov did you change the test in that last push?
i fixed log messages, pytest mark and wait_others=1 to restart methods in places not related to remove ndoe

kbr-scylla commented:
@aleksbykov I opened issue for the gossiper failure detector thing: https://github.com/scylladb/scylladb/issues/18257

I don't think we'll fix it soon, it's not critical.

In the test you should be able to work around the problem: after killing a node, wait until other nodes see it as DOWN, before restarting it.

kbr-scylla commented:
@aleksbykov the other failure you posted, the problem was that removenode was started before nodes even noticed that the killed coordinator is DOWN.

Then they marked is as DOWN, while removenode was already doing repair, and later as UP. The order of marking was correct in this case.

So in this case there is no bug. It takes time for nodes to mark dead nodes as DOWN, and then it takes time to mark restarted node as UP.

The test needs to be fixed either way: after killing a node, wait for other nodes to mark it as DOWN (you can use `server_not_sees_other_server` function). Then restart it, and wait for servers to see each other, before doing removenode.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8013/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8013/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9006/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/981/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8013/)
#### Build Details:
- Duration: 4 hr 30 min
- Builder: i-090ffa44ea7543272 (m5ad.12xlarge)

kbr-scylla commented:
https://github.com/scylladb/scylladb/pull/16197#issuecomment-2059709130

@aleksbykov please report new issue with this failure and attach logs -- looks like we don't have existing issue about this

aleksbykov commented:
> [#16197 (comment)](https://github.com/scylladb/scylladb/pull/16197#issuecomment-2059709130)
> 
> @aleksbykov please report new issue with this failure and attach logs -- looks like we don't have existing issue about this

issue: https://github.com/scylladb/scylladb/issues/18278 was opened

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8023/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8023/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/994/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9018/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8023/)
#### Build Details:
- Duration: 2 hr 33 min
- Builder: spider6.cloudius-systems.com

kbr-scylla commented:
Qd
